pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+	}
+
+	log.Info(fmt.Sprintf("found out secret %s", tokenSecret))
+
+	secret := &corev1.Secret{}
+
+	err = c.Get(context.TODO(), types.NamespacedName{"openshift-monitoring", tokenSecret}, secret)
+	if err != nil {
+		return false, err
+	}
+
+	token := secret.Data[corev1.ServiceAccountTokenKey]
+
+	route := &routev1.Route{}
+	err = c.Get(context.TODO(), types.NamespacedName{Namespace: "openshift-monitoring", Name: "prometheus-k8s"}, route)
+	if err != nil {
+		return false, err
+	}
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+
+	hclient := http.Client{Transport: tr}
+	url := "https://" + route.Spec.Host + "/api/v1/query"
+
+	req, err := http.NewRequest("GET", url, nil)
+	q := req.URL.Query()
+	alertQuery := "ALERTS{alertstate=\"firing\",severity=\"critical\",namespace=~\"^openshift.*|^kube.*|^default$\",namespace!=\"openshift-customer-monitoring\",alertname!=\"ClusterUpgradingSRE\",alertname!=\"DNSErrors05MinSRE\",alertname!=\"MetricsClientSendFailingSRE\"}"
+	q.Add("query", alertQuery)
+
+	req.URL.RawQuery = q.Encode()
+	req.Header.Add("Authorization", "Bearer "+string(token))
+	resp, err := hclient.Do(req)
+	if err != nil {
+		return false, err
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+
+	log.Info(fmt.Sprintf("alerts : %s", body))
+	alerts := &AlertResponse{}
+
+	err = json.Unmarshal(body, alerts)
+	if err != nil {
+		return false, err
+	}
+
+	if len(alerts.Data.Result) > 0 {
+		log.Info("there are critical alerts exists, cannot upgrade now")
+		//TODO send out upgrade alerts
+		return false, fmt.Errorf("there are %d critical alerts", len(alerts.Data.Result))
+	}
+
+	//check co status
+
+	operatorList := &configv1.ClusterOperatorList{}
+	err = c.List(context.TODO(), operatorList, []client.ListOption{}...)
+	if err != nil {
+		return false, err
+	}
+
+	degradedOperators := []string{}
+	for _, co := range operatorList.Items {
+		for _, condition := range co.Status.Conditions {
+			if (condition.Type == configv1.OperatorDegraded && condition.Status == configv1.ConditionTrue) || (condition.Type == configv1.OperatorAvailable && condition.Status == configv1.ConditionFalse) {
+				degradedOperators = append(degradedOperators, co.Name)
+			}
+		}
+	}
+
+	if len(degradedOperators) > 0 {
+		log.Info(fmt.Sprintf("degraded operators :%s", strings.Join(degradedOperators, ",")))
+		return false, fmt.Errorf("degraded operators :%s", strings.Join(degradedOperators, ","))
+	}
+	return true, nil
+
+}
+
+type AlertResponse struct {
+	Status string    `json:"status"`
+	Data   AlertData `json:"data"`
+}
+
+type AlertData struct {
+	Result []interface{} `json:"result"`
+}
+
+// validateUpgradeConfig will validate the UpgradeConfig, the desired version should be grater than or equal to the current version
+func ValidateUpgradeConfig(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (result bool, err error) {
```

Comment:
@mrbarge @dofinn @wanghaoran1988  Given all our other validation is in webhooks should this be moved there as well?

Reply:
@jroche:

- Re: checking daemonsets/replicasets - originally I was thinking that it's not necessary, however we still do have some things sitting in the openshift* namespace that aren't managed by operators (webhook server). If everything was a cluster operator I would agree that it's superfluous.

- Re: validate UpgradeConfig via webhooks - completely agree, that is a perfect scenario for it.

- Re: terminology of 'control plane', if we go by this page [0] then I would say that maaaybe they are part of it?

[0] https://docs.openshift.com/container-platform/4.4/architecture/control-plane.html

Reply:
https://issues.redhat.com/browse/OSD-3915 created